projects/krasnor/proposal/README.md

```diff
@@ -0,0 +1,53 @@
+
+## Motivation
+
+During the lecture we were introduced into the halfedge structure and various remeshing algorithms. In the exercises we got to know a down-sampling algorithm (QEM) much closer by implementing it, but we did not implement an up-sampling algorithm. Since I like the idea of altering the mesh structure via an algorithm I want to implement/extend the halfegde structure by an subdivision algorithm (Catmull-Clark). 
+
+## Proposal
+
+![](assets/subdivision_example.jpg)
+
+​		*A cube (left) subdivided with the Catmull-Clark subdivision algorithm after 1 (middle) and 2 (right) iterations. Example rendered with [Blender](https://www.blender.org/).*
+
+I want to implement the Catmull-Clark subdivision algorithm [Catmull et al. 1978] for a halfedge-mesh structure [Kettner 2004]. During initial research I noticed that the algorithm can convert triangle faces into quadrilateral faces. Thus the halfedge structure must have quadrilateral face support. Additionally the visualization will support the user to compare the original mesh and subdivided mesh (along with additional information, such as vertex count) by displaying both versions at the same time next to each other. As remeshing is part of geometry processing this project is relevant to the lecture. 
+
+I will use HTML5, Javascript, [Node.js](https://nodejs.org/) and the library [three.js](http://threejs.org/) for this project. 
+
+The implementation of the halfedge structure will be based on the already known code from the exercises and lecture, but will be extended to be able to handle quadrilateral faces. The Catmull-Clark subdivision algorithm will be based on the original paper [Catmull et al. 1978], but be adjusted to use the halfedge structure instead of a matrix representation for the mesh. While the 3D-modelling software [Blender](https://www.blender.org/) also implements the Catmull-Clark subdivision algorithm with a halfedge structure it is written in C++, my implementation will be written in JavaScript and three.js. Thus the project is not based on an already existing code solution.
+
+## Implementation
+
+The project will be implemented by the usage of HTML5, JavaScript, [Node.js](https://nodejs.org/) and [three.js](http://threejs.org/). 
+
+Milestones will be:
+
+1. Implementing halfedge connectivity to support quadrilateral faces
+2. Adjust calculations for normals
+3. Get three.js to handle the quadrilateral Mesh (as three.js dropped Face4 support)
+4. Implement the Catmull–Clark Subdivision algorithm
+5. Add support to compare the original mesh and subdivided mesh, along with additional information (such as vertex count)
+
+Bigger challenges may encountered in Milestones 1, 3 and 4:
+
+- Milestone 1: a lot of code has to be adjusted to enable the support for quadrilateral faces, checking validity and debugging the generated mesh will be difficult.
```

Comment:
The nice thing about halfedge representation is that a face does not require to be a triangle.
Looping the edges of a face would be as same as how it is implemented in a triangle mesh.

Probably the only change that you need to deal with is the loader where the way you construct a face.
Maintaining the connectivities seems to be the same(?).

Lastly, it is worth mentioning that a face does not have to be flat.